# 2 УРОК ФОРМУЛА БЕРНУЛЛИ

Продолжим решать задачки по теории вероятностей. В этом степе попрактикуемся в решении задач по формуле Бернулли.

Чтобы подсчитать задачу через Python, вам может пригодиться следующее:

In [2]:
from scipy import special #модуль питона, предназначенный для научных и инженерных расчетов

# special.comb(n, k) #биномиальный коэффициент



#альтернативный вариант

from math import factorial

def num_of_successes(n, k):
    return factorial(n)/(factorial(k) * factorial(n - k))

## 2

## Задание 1

Устройство, состоящее из пяти независимо работающих элементов, включается за время Т. Вероятность отказа каждого из них за это время равна 0,4. Найти вероятность того, что откажут три элемента. Ответ округлите до четырёх знаков после запятой. В качестве десятичного разделителя используйте точку.

Подсказки

- **Что надо иметь в виду при решении?**
    
    Здесь нет никаких подвохов - достаточно подставить всё необходимое в формулу Бернулли:
    
    $$
    P_{n}^{k}=C_{n}^{k}p^{k}q^{n-k}
    $$
    
    где 
    
    $$
    C_{n}^{k} = {n! \over k!(n-k)!}
    $$
    
    n в данном случае - количество элементов, k - количество тех элементов, которые должны отказать, p = вероятность отказа, q - вероятность того, что элемент будет работать (1 - p).

In [5]:
round(special.comb(5, 3)*(0.4**3)*(0.6**2),4)

0.2304

##### ОТВЕТ: 0.2304

**Разбор решения**

В данном случае мы можем вспомнить формулу Бернулли:

$$
P_{n}^{k}=C_{n}^{k}p^{k}q^{n-k}
$$

где 

$$
C_{n}^{k} = {n! \over k!(n-k)!}
$$

Тогда мы получаем, что p (вероятность отказа элемента) = 0.4, а число испытаний (т.е. число элементов) n = 5.

Вероятность того, что откажут ровно три элемента из пяти:

$$
P_{5}(3) = C_{5}^3 \times 0.4^3 \times 0.6^2 \approx 0.2304
$$

## 3

## Задание 2

Устройство, состоящее из пяти независимо работающих элементов, включается за время Т. Вероятность отказа каждого из них за это время равна 0,4. Найдите вероятность того, что хотя бы один элемент откажет. Ответ округлите до трёх знаков после запятой. В качестве десятичного разделителя используйте точку.

Подсказки

- **Что стоит учесть здесь?**
    
    Ситуация, противоположная "хоть один элемент отказал" - "ни один элемент не отказал". Если сумеете найти вероятность второй ситуации - автоматически найдёте вероятность первой.
    
    Это можно решить как средствами той же формулы Бернулли, так и простым перемножением вероятностей в духе прошлого урока.

In [27]:
round((1- 0.6**5), 3)

0.922

In [26]:
round(1-(special.comb(5, 0)*(0.4**0)*(0.6**5)),3)

0.922

##### ОТВЕТ: 0.922

В данном случае мы можем вспомнить формулу Бернулли:

$$
P_{n}^{k}=C_{n}^{k}p^{k}q^{n-k}
$$

где

$$
C_{n}^{k} = {n! \over k!(n-k)!}
$$

Тогда мы получаем, что p (вероятность отказа элемента) = 0.4, а число испытаний (т.е. число элементов) n = 5.

Проще всего рассматривать эту ситуацию с противоположной ситуации - не откажет ни один элемент.

Вероятность того, что откажет один элемент:

$$
1 - P_{5}(0) = 1 - C_{5}^0 \times 0.4^0 \times 0.6^5 \approx 0.922
$$

## 4

## Задание 3

Производится 8 выстрелов по цели, в каждом из которых вероятность попадания равна 0,3. Найти вероятность того, что цель будет поражена хотя бы два раза. Ответ округлите до тысячных. В качестве десятичного разделителя используйте точку.

Подсказки

- **Что нужно понимать при решении этой задачи?**
    
    Здесь ситуация очень похожа на предыдущую, но нужно вычислить две вероятности и сложить их:
    
    - Вероятность, что по цели вообще не будет попаданий
    - Вероятность, что попадание будет только одно
    
    Сумма этих вероятностей противоположна вероятности события "попали хотя бы два раза". 
    
    **NB!** Попали хотя бы два раза - это попали два раза или больше.

In [29]:
# Вероятность общего промаха

(1-0.3)**8

0.05764800999999997

In [32]:
# Вероятность 1 попадания
special.comb(8, 1)*(0.3**1)*(0.7**7)


0.1976503199999999

In [36]:
special.comb(8, 1)

8.0

In [33]:
1 - (0.1976503199999999 + 0.05764800999999997)

0.7447016700000001

In [35]:
round((1 - (0.1976503199999999 + 0.05764800999999997)),3)

0.745

##### ОТВЕТ: 0.745

**Разбор решения**

В данном случае мы применяем формулу Бернулли:

$$
P_{n}^{k}=C_{n}^{k}p^{k}q^{n-k}
$$

где

$$
C_{n}^{k} = {n! \over k!(n-k)!}
$$

Тогда мы получаем, что p (вероятность попадания в цель) = 0.3, а число испытаний (т.е. число выстрелов) n = 8. Число попаданий в цель в данном случае будет k = 2.

Теперь рассмотрим ситуацию, что мы хотим попасть хотя бы два раза, т.е. нас не устраивают случаи, когда мы совсем не попадаем (k = 0) или попадаем 1 раз (k = 1).

Подставляем значения в формулу Бернулли:



$$
1 - C_{8}^{1} \times 0.3^{1} \times (1-0.3)^{8-1} - C_{8}^{0} \times 0.3^{0} \times (1-0.3)^{8} = 0.745
$$

## 5

## Задание 4

В магазине 7 покупателей. Каждый может совершить покупку с вероятностью 0,4. Найти вероятность того, что не более двух человек совершат покупку. Ответ округлите до сотых. В качестве десятичного разделителя используйте точку.

Подсказки

- **А что делать тут?**
    
    Обратите внимание: теперь у нас не "хотя бы", а "не более"! Иными словами, нам больше не нужно вычитать из 1. 
    
    В событие "совершили покупку не более двух раз" входят три события, вероятности которых надо вычислить и сложить:
    
    - Не совершили покупку вообще
    - Совершили одну покупку
    - Совершили две покупки

In [40]:
# Совершили две покупки
(special.comb(7, 2)*(0.4**2)*(0.6**5))


0.2612736

In [39]:
special.comb(7, 2)

21.0

In [41]:
# Не совершили покупку вообще
0.6**7


0.027993599999999993

In [43]:
# Совершили одну покупку
special.comb(7, 1)*(0.4**1)*(0.6**6)

0.13063679999999997

In [44]:
round((0.2612736+0.027993599999999993+0.13063679999999997),2)

0.42

##### ОТВЕТ: 0.42

**Разбор решения**

В данном случае мы применяем формулу Бернулли:

$$
P_{n}^{k}=C_{n}^{k}p^{k}q^{n-k}
$$

где

$$
C_{n}^{k} = {n! \over k!(n-k)!}
$$

Тогда мы получаем, что p (вероятность совершения покупки) = 0.4, а число испытаний (т.е. число покупателей) n = 7.

Нас здесь интересуют три варианта: когда никто не совершил покупку (k = 0), когда её совершил только один покупатель (k = 1) и когда её совершило два покупателя (k = 2).

Подставляем значения в формулу Бернулли:


C_{7}^{0} \times 0.4^{0} \times (1-0.4)^{7} + C_{7}^{1} \times 0.4^{1} \times (1-0.4)^{7-1} + C_{7}^{2} \times 0.4^{2} \times (1-0.4)^{7-2} \approx 0.42

$$
C_{7}^{0} \times 0.4^{0} \times (1-0.4)^{7} + C_{7}^{1} \times 0.4^{1} \times (1-0.4)^{7-1} + C_{7}^{2} \times 0.4^{2} \times (1-0.4)^{7-2} \approx 0.42
$$

# 6

## Задание 5

В телеателье имеется 7 телевизоров. Для каждого телевизора вероятность того, что в данный момент он включен, равна 0,5. Найти вероятность того, что в данный момент включены не менее трех телевизоров. Ответ округлите до тысячных. В качестве десятичного разделителя используйте точку.

Подсказки

- **А здесь что?**
    
    Ситуация почти такая же, как и в прошлой задаче, но теперь написано "не менее". Это словосочетание эквивалентно "хотя бы". Иными словами:
    
    - Теперь мы опять вычитаем из 1
    - Вероятность надо поменять на 0.5

In [49]:
# все выключены
0.5**7

0.0078125

In [52]:
# включен 1
(1*2*3*4*5*6*7/(1 * 1*2*3*4*5*6)) * 0.5**1 * 0.5**6

0.0546875

In [53]:
# включен 2
special.comb(7, 2)*(0.5**2)*(0.5**5)

0.1640625

In [56]:
# Остальные  варианты нам подходят

round(1-(0.1640625+0.0546875+0.0078125),3)

0.773

##### ОТВЕТ: 0.773

**Разбор решения**

$$
В \ данном \ случае \ мы \ применяем \ формулу \ Бернулли: P_{n}^{k}=C_{n}^{k}p^{k}q^{n-k}, 
$$

$$
C_{n}^{k} = {n! \over k!(n-k)!}
$$

Тогда мы получаем, что p (вероятность того, что телевизор включен) = 0.5, а число испытаний (т.е. число телевизоров) n = 7.

Здесь нам не подходят только три варианта: когда не включён ни один телевизор (k=0), когда включён только один (k=1) и когда включены только два (k=2). Соответственно, через обратную вероятность мы сможем вычислить всё остальное.

Подставляем значения в формулу Бернулли:

$$
1 - C_{7}^{0} \times 0.5^{0} \times (1-0.5)^{7} - C_{7}^{1} \times 0.5 \times (1-0.5)^{7-1} - C_{7}^{2} \times 0.5^{2} \times (1-0.5)^{7-2} = 0.773
$$

## 7

## Задание 6

Пусть вероятность того, что покупателю необходим компьютер с операционной системой Linux, равна 0.25. Найдите вероятность того, что из шести покупателей по крайней мере двум необходим компьютер с операционной системой Linux.

Ответ округлите до сотых, в качестве разделителя используйте точку.

Если нужна помощь, пишите сюда

Подсказки

- **А здесь что?**
    
    Идея всё та же: "по крайней мере двум" означает "не один и не ноль". Соответственно, вам надо посчитать вероятности того, что Linux никому не нужен и что он нужен только 1 человеку из шести. Эти две вероятности вычитаете из 1 и получаете нужное число.

In [61]:
# Никому не нужен
(1-0.25)**6

0.177978515625

In [62]:
# нужен  только 1-му
special.comb(6, 1)*(0.25**1)*(0.75**5)

0.35595703125

In [64]:
round((1-(0.177978515625+0.35595703125)), 2)

0.47

##### ОТВЕТ: 0.47

**Разбор решения**

Воспользуемся понятием отрицательной вероятности, т. е. P(A) = 1 - P(не А)

В нашем случае, A - такое событие, что из шести покупателей по крайней мере двум необходим компьютер с операционной системой Linux

Соответственно, не А - такое событие, что одному покупателю необходим компьютер или ни одному покупателю не необходим компьютер. В таком случае:

1 - (P(одному покупателю необходим компьютер c Linux) + P(никому не необходим компьютер с Linux)) = 

$$
1 - C_{6}^{1}0.25^{1}0.75^{5} + C_{6}^{0}0.25^{0}0.75^{6} ≈ 0.47
$$

## 8

## Задание 7

Для нормальной работы отдела по дата-аналитике необходимо, чтобы в офисе было не менее восьми программистов, а всего в штате 10 программистов. Вероятность того, что программист заболеет, равна 0.1. Найдите вероятность нормальной работы отдела по дата-аналитике.

Ответ округлите до сотых, а в качестве разделителя используйте точку.

Подсказки 

- **А здесь как быть?**
    
    Отдел будет работать нормально в трёх случаях:
    
    - Не болеет никто
    - Болеет один программист
    - Болеют двое программистов
    
    Суммируете вероятности этих событий и получаете нужное число.

In [68]:
# Надо найти вероятность что заболеют 2 чел и сложить
special.comb(10, 2)*(0.1**2)*(0.9**8)

0.19371024450000007

In [70]:
# Надо найти вероятность что заболеют 1 чел и сложить
special.comb(10, 1)*(0.1**1)*(0.9**9)

0.3874204890000001

In [1]:
# Надо найти вероятность что не заболеют никто
0.9**10

0.3486784401000001

In [74]:
round((0.19371024450000007+0.3486784401000001+0.3874204890000001), 2)

0.93

##### ОТВЕТ: 0.93

**Разбор решения**

В отличии от предыдущей задачи, тут нам удобнее посчитать вероятности того, что в офисе 8, 9 или 10 программистов. Это можно сделать как:

$$
P = C_{10}^{8}0.9^{8}0.1^{2} + C_{10}^{9}0.9^{9}0.1^{1} + C_{10}^{10}0.9^{10}0.1^{0} = 0.93
$$

## 9

## Задание 8

Среди всех скриптов, которые пишет программист, с 10%-ой вероятностью скрипт будет с ошибкой. Найдем вероятность того, что среди 30 скриптов два не будут работать.

Ответ округлите до тысячных, в качестве десятичного разделителя используйте точку.

Подсказки

## - **Что делаем тут?**
    
    Здесь вообще без подвохов - просто считаете по формуле Бернулли вероятность того, что 2 скрипта из 30 не будут работать.

In [79]:
round((special.comb(30, 2)*(0.1**2)*(0.9**28)), 3)

0.228

###### ОТВЕТ : 0.228

**Разбор решения**

Достаточно просто по схеме Бернулли посчитать вероятность того, что два скрипта не будут работать

$$
P = C_{30}^{2}0.1^{2}0.9^{28} ≈ 0.228
$$

## 10

## Задание 9

В очень большом чате по дата аналитике сидит 1730 человек. Вероятность того, что день рождения наугад взятого участника чата приходится на определенный день года, равна 1/365 для каждого из 365 дней. Найдем наиболее вероятное число счастливчиков, которые родились 1 марта.

Для решения этой задачи вам может потребоваться эта формула.

Подсказки

Биномиальное распределение (распределение по схеме Бернулли) позволяет, в частности, установить, какое число появлений события А наиболее вероятно. Формула для наиболее вероятного числа успехов k (появлений события) имеет вид:

np−q≤k≤np+p,q=1−p.
Так как np−q=np+p−1, то эти границы отличаются на 1. Поэтому k, являющееся целым числом, может принимать либо одно значение, когда np целое число (k=np) , то есть когда np+p (а отсюда и np−q) нецелое число, либо два значения, когда np−q целое число.

- **А это ещё что такое?**
    
    А тут в дело вступает новая формула! Она называется **формулой наивероятнейшего числа успехов** и приложена к заданию по [вот этой ссылке](https://www.matburo.ru/tvart_sub.php?p=calc_bern_naiv).
    
    Сама формула выглядит так:
    
    $$
    np - q \leq m \leq np + p
    $$
    
    В нашем случае $n$ - количество людей в чатике, $p$ - вероятность, что у человека день рождения, $q$ - вероятность, что у человека не день рождения, $m$ - интересующее нас число (должно быть целым).

In [80]:
1730*(1/365) - 364/365

3.742465753424658

In [83]:
1730*(1/365) + 1/365

4.742465753424658

In [ ]:
ОТВЕТ: 4

**Разбор решения**

В нашем случае,  n = 1730, p = 1/365, q = 364/365

$$
Тогда, np - q \leq k \leq np + p
$$

$$
3.742 \leq k \leq 4.742
$$

k = 4

Ответ: 4

## 11

## Задание 10

В урне 20 шаров: 15 белых и 5 чёрных. Вынули подряд 5 шаров, причём каждый вынутый шар возвращается в урну и перед извлечением следующего шары в урне тщательно перемешиваются. Найдите вероятность того, что из пяти вынутых шаров будет 2 белых.

Ответ округлите до тысячных, в качестве десятичного разделителя используйте точку.

Подсказки

- **Неужели здесь просто формула Бернулли?**
    
    Она самая. Вероятности событий считаете как количество шаров определённого цвета относительно общего числа шаров, а потом считаете вероятность 2 белых шаров из 5 выбранных. 
    
    Так как шары убираются обратно, вероятность их достать никак не меняется.

In [84]:
# вероятность белого 
15/20

0.75

In [85]:
# Вероятность черного
5/20

0.25

In [86]:
round((special.comb(5, 2)*(0.75**2)*(0.25**3)), 3)

0.088

###### ОТВЕТ: 0.088

**Разбор решения**

Вероятность достать белый шарик всегда равна 15/20 = 0.75, так как все шары возвращаются в урну. Соответственно, шанс достать черный шар всегда 1/4 = 0.25

По схеме Бернулли:

$$
C_{5}^{2} * 0.75 ^ 2 * 0.25 ^ 3 ≈ 0.088
$$